Bluetooth: GATT: Add support to setting permission on CCCD

Vudentz opened the pull request:
This adds support to set different permissions to CCCD so security can
be checked when enabling notification which conforms to:

BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2360:

  '3.3.3.3 Client Characteristic Configuration

   Authentication and authorization may be required by the server to
   write the configuration descriptor.'

In addition to that also ensure that notification are not re-enabled
until the proper security level is reached to conform to the following
statement:

  '10.3.1.1 Handling of GATT indications and notifications

   A client “requests” a server to send indications and notifications
   by appropriately configuring the server via a Client Characteristic
   Configuration Descriptor. Since the configuration is persistent
   across a disconnection and reconnection, security requirements must
   be checked against the configuration upon a reconnection before
   sending indications or notifications. When a server reconnects to a
   client to send an indication or notification for which security is
   required, the server shall initiate or request encryption with the
   client prior to sending an indication or notification. If the client
   does not have an LTK indicating that the client has lost the bond,
   enabling encryption will fail.'

Fixes #17983

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Add support to setting permission on CCCD

This adds support to set different permissions to CCCD so security can
be checked when enabling notification which conforms to:

BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2360:

  '3.3.3.3 Client Characteristic Configuration

   Authentication and authorization may be required by the server to
   write the configuration descriptor.'

In addition to that also ensure that notification are not re-enabled
until the proper security level is reached to conform to the following
statement:

  '10.3.1.1 Handling of GATT indications and notifications

   A client “requests” a server to send indications and notifications
   by appropriately configuring the server via a Client Characteristic
   Configuration Descriptor. Since the configuration is persistent
   across a disconnection and reconnection, security requirements must
   be checked against the configuration upon a reconnection before
   sending indications or notifications. When a server reconnects to a
   client to send an indication or notification for which security is
   required, the server shall initiate or request encryption with the
   client prior to sending an indication or notification. If the client
   does not have an LTK indicating that the client has lost the bond,
   enabling encryption will fail.'

Fixes #17983

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

carlescufi reviewed (commented):
Looks pretty good to me, thanks @Vudentz. Don't we need a check in `bt_gatt_notify()` though? if the application passes a `NULL` pointer we still need to check the security level at that point, or is that done in the connection callback?

Vudentz commented:
> Looks pretty good to me, thanks @Vudentz. Don't we need a check in `bt_gatt_notify()` though? if the application passes a `NULL` pointer we still need to check the security level at that point, or is that done in the connection callback?

The callback will check the value set in the user_data (bt_gatt_ccc), that will be set to 0x00 (disabled) in case it could not be restored due to security level, to restored CCC which requires security bt_gatt_encrypt_change shall be called.

nvlsianpu commented:
cc @rakons

Vudentz commented:
@jhedberg @carlescufi It seems we have forgotten this one.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1735,6 +1741,25 @@ static u8_t connected_cb(const struct bt_gatt_attr *attr, void *user_data)
+			/* Check if current security is enough */
+			if (conn->sec_level < sec) {
+				if (data->sec < sec)
+					data->sec = sec;
```
Missing `{ }`

Vudentz replied:
Fixed.

jhedberg reviewed (approved):
Looks fine except for one coding style issue (missing braces)

jhedberg commented:
@carlescufi @joerchan could you please review?

carlescufi commented:
@Vudentz @jhedberg yes, apologies for the delay on this one. We have had holidays and travelling. We will review this week.

nvlsianpu commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1735,6 +1741,26 @@ static u8_t connected_cb(const struct bt_gatt_attr *attr, void *user_data)
 
+#if defined(CONFIG_BT_SMP)
+		/* Check if attribute requires encryption/authentication */
+		if (attr->perm & BT_GATT_PERM_WRITE_ENCRYPT ||
```
```suggestion
		if (attr->perm &( BT_GATT_PERM_WRITE_ENCRYPT | BT_GATT_PERM_WRITE_AUTHEN)) {
```
reason: less assembly instruction in most cases

Vudentz replied:
Indeed that would be better, will send an update shortly.

Vudentz replied:
I did the update but found out we could reuse the logic for checking the attribute permission for the CCC, so I end up adding another patch.

nvlsianpu commented:
cc @pdunaj

Vudentz committed:
Bluetooth: GATT: Use check_perm to check CCC permission

This moves check_perm code under GAT and make use of it to check if CCC
write permission can be attended by the connection security level or if
it needs to be increased.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented on subsys/bluetooth/host/gatt_internal.h:
```diff
@@ -11,8 +11,20 @@
 #define BT_GATT_CENTRAL_ADDR_RES_NOT_SUPP	0
 #define BT_GATT_CENTRAL_ADDR_RES_SUPP		1
 
+#define BT_GATT_PERM_READ_MASK			(BT_GATT_PERM_READ | \
```
This is not a public API, it is only to be used with bt_gatt_check_perm which is used by att.c as well.

Vudentz replied:
Id like this to be merge at some point so we need to be clear if this requires another change or it is fine to be applied.

nvlsianpu replied:
Ach right - I missed _MASK at the ends

jhedberg commented:
@Vudentz the shippable failure seems valid. Didn't check all failures but this one is at least there:
```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/gatt.c: In function 'update_ccc':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/gatt.c:1767:7: error: unused variable 'err' [-Werror=unused-variable]
  u8_t err;
       ^~~
```

Vudentz commented:
> @Vudentz the shippable failure seems valid. Didn't check all failures but this one is at least there:
> 
> ```
> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/gatt.c: In function 'update_ccc':
> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/gatt.c:1767:7: error: unused variable 'err' [-Werror=unused-variable]
>   u8_t err;
>        ^~~
> ```

Yep, Ive just fixed it.

joerchan commented on samples/bluetooth/peripheral/src/main.c:
```diff
@@ -194,7 +194,8 @@ BT_GATT_SERVICE_DEFINE(vnd_svc,
 			       read_vnd, write_vnd, vnd_value),
-	BT_GATT_CCC(vnd_ccc_cfg_changed),
+	BT_GATT_CCC(vnd_ccc_cfg_changed,
+		    BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT),
```
Is this a requirement of the VND service?

Vudentz replied:
Yes it follows the chrc permissions.

joerchan commented on include/bluetooth/conn.h:
```diff
@@ -334,6 +334,12 @@ typedef enum __packed {
+ *
+ *  @return Connection security level
+ */
+bt_security_t bt_conn_get_security(struct bt_conn *conn);
```
Maybe move below the __deprecated declaration of the old bt_conn_security?

Vudentz replied:
For what reason?

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1085,6 +1085,16 @@ int bt_conn_set_security(struct bt_conn *conn, bt_security_t sec)
+#else
+bt_security_t bt_conn_get_security(struct bt_conn *conn)
+{
+	return BT_SECURITY_L1;
```
Should it be L0 for BR/EDR? (It looks like conn->sec_level is not initialized for BR/EDR connections, see `bt_conn_add_br`)

Vudentz replied:
Well L0 is only used for SDP iirc.

joerchan reviewed (approved):
Looks good. Just some minor questions.

Vudentz commented:
@jhedberg can we finally gets this merged?

jhedberg commented:
> @jhedberg can we finally gets this merged?

@Vudentz done!
